components/background_hang_monitor/sampler_linux.rs

```diff
@@ -207,59 +207,59 @@ impl Sampler for LinuxSampler {
-        let shared_state = unsafe { &*SHARED_STATE.0.get() };
-        shared_state
-            .msg2
-            .as_ref()
-            .unwrap()
-            .wait_through_intr()
-            .expect("msg2 failed");
-
-        let context = CONTEXT.load(Ordering::SeqCst);
-        let mut cursor = mem::MaybeUninit::uninit();
-        let ret = unsafe { unw_init_local(cursor.as_mut_ptr(), context) };
-        let result = if ret == UNW_ESUCCESS {
-            let mut native_stack = NativeStack::new();
-            loop {
-                let ip = match get_register(cursor.as_mut_ptr(), RegNum::Ip) {
-                    Ok(ip) => ip,
-                    Err(_) => break,
-                };
-                let sp = match get_register(cursor.as_mut_ptr(), RegNum::Sp) {
-                    Ok(sp) => sp,
-                    Err(_) => break,
-                };
-                if native_stack
-                    .process_register(ip as *mut _, sp as *mut _)
-                    .is_err() ||
-                    !step(cursor.as_mut_ptr()).unwrap_or(false)
-                {
-                    break;
+        let result;
+        {
+            let shared_state = unsafe { &*SHARED_STATE.0.get() };
+            shared_state
+                .msg2
+                .as_ref()
+                .unwrap()
+                .wait_through_intr()
+                .expect("msg2 failed");
+
+            let context = CONTEXT.load(Ordering::SeqCst);
+            let mut cursor = mem::MaybeUninit::uninit();
+            let ret = unsafe { unw_init_local(cursor.as_mut_ptr(), context) };
+            result = if ret == UNW_ESUCCESS {
+                let mut native_stack = NativeStack::new();
+                loop {
+                    let ip = match get_register(cursor.as_mut_ptr(), RegNum::Ip) {
+                        Ok(ip) => ip,
+                        Err(_) => break,
+                    };
+                    let sp = match get_register(cursor.as_mut_ptr(), RegNum::Sp) {
+                        Ok(sp) => sp,
+                        Err(_) => break,
+                    };
+                    if native_stack
+                        .process_register(ip as *mut _, sp as *mut _)
+                        .is_err() ||
+                        !step(cursor.as_mut_ptr()).unwrap_or(false)
+                    {
+                        break;
+                    }
                 }
-            }
-            Ok(native_stack)
-        } else {
-            Err(())
-        };
-
-        // signal the thread to continue.
-        shared_state
-            .msg3
-            .as_ref()
-            .unwrap()
-            .post()
-            .expect("msg3 failed");
-
-        // wait for thread to continue.
-        shared_state
-            .msg4
-            .as_ref()
-            .unwrap()
-            .wait_through_intr()
-            .expect("msg4 failed");
-
-        // No-op, but marks the end of the shared borrow
-        drop(shared_state);
```

Comment:
I think the simpler change is to just remove this line, since we know it's a no-op?

Reply:
Won't that cause a deadlock since `clear_shared_state()` also does a `SHARED_STATE.0.get()`?

Reply:
I guess not, because it's an [UnsafeCell](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html)( I tried in the playground you can call `get` multiple times without deadlock), but, better be safe than sorry and you're version show the intent of "dropping the shared borrow"... Ok merge it as is?

Reply:
Okay. That makes sense. I'll go ahead and merge.